### PR TITLE
JDK-8317372: Refactor some NumberFormat tests to use JUnit

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR which refactors a number of tests under `test/text/NumberFormat` to use JUnit.

During the switch to JUnit, the tests had the following updates (to improve readability)
- separate the test data generation from the actual execution of the test
- create distinct test methods so that all the tests aren't just run under one big method (e.g. main)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317372](https://bugs.openjdk.org/browse/JDK-8317372): Refactor some NumberFormat tests to use JUnit (**Sub-task** - P4)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16035/head:pull/16035` \
`$ git checkout pull/16035`

Update a local copy of the PR: \
`$ git checkout pull/16035` \
`$ git pull https://git.openjdk.org/jdk.git pull/16035/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16035`

View PR using the GUI difftool: \
`$ git pr show -t 16035`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16035.diff">https://git.openjdk.org/jdk/pull/16035.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16035#issuecomment-1751272410)

### CODE REVIEW

File: test/jdk/java/text/Format/NumberFormat/Bug4838107.java [hunk 1]
```diff
@@ -24,228 +24,256 @@
...
-            }
+            assertEquals(parsed2, parsed1, "DecimalFormat round trip parse(" +
+                    num.getClass().getName() +
+                    ") error: \n\toriginal number:  " + str +
+                    "\n\tparsed number:    " + parsed1 +
+                    "  (" + parsed1.getClass().getName() + ")" +
+                    "\n\tformatted number: " + formatted +
+                    "\n\tre-parsed number: " + parsed2 +
+                    "  (" + parsed2.getClass().getName() + ")" +
+                    "\n\tminus sign: " + df.getDecimalFormatSymbols().getMinusSign());
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/Bug4838107.java:254]: Could use text block to get rid of `\n`/`\t`s, with `.formatted()`. Same for the following one.

File: test/jdk/java/text/Format/NumberFormat/Bug4944439.java [hunk 2]
```diff
@@ -25,87 +25,95 @@
...
+        Locale.setDefault(savedLocale);
+    }
 
-        Locale.setDefault(defaultLoc);
+    // Check return type and value returned by DecimalFormat.parse() for longs
+    @ParameterizedTest
+    @MethodSource("longs")
+    public void parseLongTest(String s) {
+        Number parsedNumber = assertDoesNotThrow(() -> df.parse(s),
+                "DecimalFormat.parse(\"%s\") should not throw ParseException");
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/Bug4944439.java:71]: `ParseException` -> `Exception`. Applies to other locations

File: test/jdk/java/text/Format/NumberFormat/Bug4990596.java [hunk 3]
```diff
@@ -21,20 +21,32 @@
...
 public class Bug4990596 {
 
-    public static void main(String[] args) {
-        new DecimalFormat().format(new MutableInteger(0));
+    // Test that a custom subclass of Number can be formatted by
+    // DecimalFormat without throwing an IllegalArgumentException
+    @Test
+    public void formatSubclassedNumberTest() {
+        assertDoesNotThrow(() -> new DecimalFormat().format(new MutableInteger(0)),
+                "DecimalFormat.parse() should support subclasses of Number");
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/Bug4990596.java:45]: DecimalFormat.format()?

File: test/jdk/java/text/Format/NumberFormat/Bug8132125.java [hunk 4]
```diff
@@ -26,20 +26,27 @@
...
-    public static void main(String[] args) {
+
+    // Ensure the CLDRConverter does not omit the Swiss number elements
+    @Test
+    public void swissNumElementsTest() {
         Locale deCH = Locale.of("de", "CH");
         NumberFormat nf = NumberFormat.getInstance(deCH);
 
-        String expected = "54\u2019839\u2019483.142"; // i.e. "\u2019" as decimal separator, "\u2019" as grouping separator
+        // i.e. "\u2019" as decimal separator, "\u2019" as grouping separator
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/Bug8132125.java:47]: This is incorrect (sorry, I am to be blamed as I wrote this 😄). Should be "`\u002E` as decimal separator, " also remove `i.e.` if you move the comment on top of the variable declaration

File: test/jdk/java/text/Format/NumberFormat/CurrencyFormat.java [hunk 5]
```diff
@@ -42,127 +46,140 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
 public class CurrencyFormat {
 
-    private static boolean isCompat;
+    // Expected data is switched depending on COMPAT or CLDR
+    // currencySymbolsTest() is only ran for COMPAT
+    private static final boolean isCompat =
+            "COMPAT".equals(System.getProperty("java.locale.providers"));
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/CurrencyFormat.java:62]: This check could be moved into `currencySymbolsTest()`

File: test/jdk/java/text/Format/NumberFormat/BigDecimalCompatibilityTest.java [hunk 6]
```diff
@@ -84,51 +111,69 @@ static private void testBigDecimal() {
...
-            num = df.parse(s);
-        }
-        catch (ParseException e) {
-            err = true;
-            System.err.println("Failed: Exception occurred: " + e.getMessage());
-            return;
-        }
+        // Check parse and returned value
+        Number parsedValue = assertDoesNotThrow(()-> df.parse(longString),
+                "Should not throw a ParseException");
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/text/Format/NumberFormat/BigDecimalCompatibilityTest.java]: Thank you for the review. I Adjusted the message, but clarified in the comment that the original intention was checking for a `ParseException`; fixed in the other instances as well.

File: test/jdk/java/text/Format/NumberFormat/Bug4838107.java (hunk 1 above)
Comment (justin-curtis-lu) [reply test/jdk/java/text/Format/NumberFormat/Bug4838107.java]: Adjusted.

File: test/jdk/java/text/Format/NumberFormat/Bug8132125.java (hunk 4 above)
Comment (justin-curtis-lu) [reply test/jdk/java/text/Format/NumberFormat/Bug8132125.java]: Fixed.

File: test/jdk/java/text/Format/NumberFormat/CurrencyFormat.java (hunk 5 above)
Comment (justin-curtis-lu) [reply test/jdk/java/text/Format/NumberFormat/CurrencyFormat.java:62]: I left it as a static declaration as although it is used in `currencySymbolsTest()`, it is also used in the data provider of `CurrencyFormatTest()`.

Comment (naotoj) [reply test/jdk/java/text/Format/NumberFormat/CurrencyFormat.java:62]: Missed that. Thanks.

File: test/jdk/java/text/Format/NumberFormat/Bug8132125.java [hunk 7]
```diff
@@ -26,20 +26,27 @@
...
-    public static void main(String[] args) {
+
+    // Ensure the CLDRConverter does not omit the Swiss number elements
+    @Test
+    public void swissNumElementsTest() {
         Locale deCH = Locale.of("de", "CH");
         NumberFormat nf = NumberFormat.getInstance(deCH);
 
-        String expected = "54\u2019839\u2019483.142"; // i.e. "\u2019" as decimal separator, "\u2019" as grouping separator
+        // "\u2019" as decimal separator, "\u002E" as grouping separator
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/Bug8132125.java:47]: "\u002E" is decimal, "\u2019" is grouping.

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**